E2E: Product Query: Popular filters presets

dinhtungdu opened the pull request:
Fixes #7933

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

This PR adds E2E tests to verify the functionalities of Popular filters presets. The failed test is related to Gutenberg 14.8.0. The Popular Filter tests are passing: https://github.com/woocommerce/woocommerce-blocks/actions/runs/3805026894/jobs/6472662871#step:12:241

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

To run the Gutenberg only test locally:
1. Change the directory to the plugin root directory in the terminal.
2. Destroy the current `wp-env` instance: `npm run wp-env destroy`
3. Setup Gutenberg config for `wp-env`: `node bin/wp-env-with-gutenberg.js`
4. Spin up new test environment: `npm run wp-env start`
5. Run the test with `GUTENBERG_EDITOR_CONTEXT` set to `gutenberg`: `GUTENBERG_EDITOR_CONTEXT=gutenberg npm run test:e2e tests/e2e/specs/backend/product-query/popular-filters.test.ts`

dinhtungdu committed:
wip: start working on popular filters

dinhtungdu committed:
test default preset

dinhtungdu committed:
Merge branch 'trunk' into tests/popular-filters

dinhtungdu committed:
Merge branch 'trunk' into tests/popular-filters

dinhtungdu committed:
test file clean up

dinhtungdu committed:
test products displayed in editor preview and block front end

dinhtungdu committed:
give editor time to update after resetting content

dinhtungdu committed:
restore the test name

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
try: fixing flaky test by waiting for suggestion items

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
use block.name when possible

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
update atomic tests to reflect changes introduced in #7815

dinhtungdu committed:
wait for elements before assertion

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
update rating test as the default template removed rating block

dinhtungdu committed:
fix rating test

dinhtungdu committed:
fix advanced filters flaky test

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
migrate improvements from #7719

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
ensure the setting is applied in the editor before saving and checking on the front end

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
Merge branch 'trunk' into fix/flaky-test-interact-with-form-token-field

dinhtungdu committed:
Merge branch 'fix/flaky-test-interact-with-form-token-field' into tests/popular-filters

dinhtungdu committed:
import const and utilities from common.ts

dinhtungdu committed:
test sorted by title preset

dinhtungdu committed:
extract common logic to utilites, Newest preset test

dinhtungdu committed:
best selling and top rated test

dinhtungdu committed:
Merge branch 'trunk' into tests/popular-filters

dinhtungdu committed:
replace openBlockEditorSettings by ensureSidebarOpened

dinhtungdu committed:
fix flakiness of setting preset

dinhtungdu committed:
fix: shorter and faster way to reset the product query page

dinhtungdu committed:
Merge branch 'trunk' into tests/popular-filters

dinhtungdu committed:
try: not using beforeEeach

dinhtungdu commented:
@gigitux FYI this will create conflicts to your PR https://github.com/woocommerce/woocommerce-blocks/pull/7956:
- The local constants and utilities are exported from `common.ts` as I mentioned in our chat.
- `getFormElementIdByLabel` gets an update to accept class with or without leading `.`.

dinhtungdu committed:
Revert "try: not using beforeEeach"

This reverts commit 7259f2d192a3236b57ce16781f0578dbe89a0b33.

dinhtungdu committed:
try: timeout

dinhtungdu committed:
use timeout after selecting the preset, giving the editor time to update

dinhtungdu committed:
verify the popular filter selection

dinhtungdu committed:
Merge branch 'trunk' into tests/popular-filters

dinhtungdu committed:
Merge branch 'trunk' into tests/popular-filters

sunyatasattva commented on tests/e2e/specs/backend/product-query/advanced-filters.test.ts:
```diff
@@ -0,0 +1,160 @@
+			} );
+
+			it( 'All statuses are enabled by default', async () => {
+				await expect( $productFiltersPanel ).toMatch( 'In stock' );
+				await expect( $productFiltersPanel ).toMatch( 'Out of stock' );
+				await expect( $productFiltersPanel ).toMatch( 'On backorder' );
```
Not a big deal, but should this also be perhaps related to the fixture data?

We could have a loop on something like `getFixtureProductsData( 'stock_statii' )` and make it more future proof?

dinhtungdu replied:
- Fixture doesn't have any information about all available statuses right now, we need to hard-code it if we want to get stock statuses from it.
- In the future, if WC Core changes the default stock statuses, we will need to update the fixture as well. To me, updating this test or updating the fixture is similar. Test should be updated as the WC and WC Blocks update.
- I checked WooCommerce REST API to see if there is any endpoint returning available stock status, but sadly there is none.
- We can centeralize the reference for available stock status options, but it should be a utility or a constants, not in the fixture data.

sunyatasattva commented on tests/e2e/specs/backend/product-query/popular-filters.test.ts:
```diff
@@ -0,0 +1,127 @@
+			it( 'Products are displayed in the correct order', async () => {
+				const { productQueryProducts, shortcodeProducts } =
+					await setupProductQueryShortcodeComparison(
+						'[products orderby="title" order="ASC" limit="9"]'
```
I would refactor these shortcodes in a variable above, to make them less buried and easier to spot.

dinhtungdu replied:
I addressed this and your other comments in [e60cf03](https://github.com/woocommerce/woocommerce-blocks/pull/7749/commits/e60cf032c9328f974822324b864e64ba59a49db3)

sunyatasattva commented on tests/e2e/specs/backend/product-query/popular-filters.test.ts:
```diff
@@ -0,0 +1,127 @@
+				expect( currentFilter ).toEqual( 'Sorted by title' );
+			} );
+
+			it( 'Editor preview and block frontend display the same products', async () => {
+				const { previewProducts, frontEndProducts } =
+					await setupEditorFrontendComparison();
+				expect( frontEndProducts ).toEqual( previewProducts );
+			} );
```
This seems to be repeated in every test case, why not just refactoring it to avoid duplicates?

sunyatasattva commented on tests/e2e/specs/backend/product-query/popular-filters.test.ts:
```diff
@@ -0,0 +1,127 @@
+				expect( frontEndProducts ).toEqual( previewProducts );
+			} );
+
+			it( 'Products are displayed in the correct order', async () => {
+				const { productQueryProducts, shortcodeProducts } =
+					await setupProductQueryShortcodeComparison(
+						'[products best_selling="true" limit="9"]'
+					);
+				expect( productQueryProducts ).toEqual( shortcodeProducts );
+			} );
```
This is also repeated over and over, the only thing that's changing is the shortcode. So you could refactor it into a function that accepts a shortcode string and that would save lines of code and make it more maintainable.

sunyatasattva reviewed (changes requested):
I've got very little to comment about this PR: this is excellent work Tung!

nefeline commented on tests/e2e/specs/backend/product-query/advanced-filters.test.ts:
```diff
@@ -0,0 +1,160 @@
+					'Show only products on sale'
+				);
+				await toggleAdvancedFilter( 'Sale status' );
+				await expect( $productFiltersPanel ).not.toMatch(
```
What do you think about reusing more code here by extracting common/duplicated lines into utility functions to make the tests easier to read and maintain?

dinhtungdu replied:
Can you suggest something here? At the moment, I don't have a good idea to refactor this further.

From the test code I can see the test perform these step:

1. Toggle the Sale status filter.
2. Verify the panel contains the option label.
3. Toggle the Sale status filter again.
4. verify the panel doesn't contain the option label anymore.

I'm curious about what you read from the test. If we can do anything to increase the readability, I'm all in.

nefeline replied:
I was thinking something along these lines (for the editor preview test, for example):

```
const toggleSaleStatusFilter = async () => {
	await toggleAdvancedFilter( 'Sale status' );
	await setCheckbox( await getToggleIdByLabel( 'Show only products on sale' ) );
};

const removeSaleStatusFilter = async () => {
	await toggleAdvancedFilter( 'Sale status' );
	await unsetCheckbox( await getToggleIdByLabel( 'Show only products on sale' ) );
};
```

But that is super minor/a nit! The tests are now structured nicely, a-ok to keep them as-is :)

nefeline commented on tests/e2e/specs/backend/product-query/advanced-filters.test.ts:
```diff
@@ -0,0 +1,160 @@
+				);
+				expect( await getPreviewProducts() ).toHaveLength(
+					defaultCount
+				);
```
I would recommend separating these into individual tests: one for when toggling on the sale status filter, and another for when toggling off, so we have a better flow and separation of concerns. Something along these lines:

```
Toggle on the 'Show only products on sale' filter
- Test the expected behavior in the editor
- Test the expected behavior on the FE

Toggle off the 'Show only products on sale' filter
- Test the expected behavior in the editor
- Test the expected behavior on the FE
```

dinhtungdu replied:
This is a great direction, thank you so much, I ended up structuring tests like following:
```
- Enable Sale Status > Editor preview shows only on sale products
- Enable Sale Status > On the front end, block shows only on sale products
```

I moved the tests for the default state to the main `describe` block to avoid duplicated tests. Same for your review about stock status.

Some notes:
- I use `>` in the test name instead of creating another nested `describe` block to save one or two indentation levels, too many nested `describe` blocks make the test harder to read.
- I explicitly mention the expected behavior in the test title.

See [8afad60](https://github.com/woocommerce/woocommerce-blocks/pull/7749/commits/8afad60be424aabc3361928d62bbea5c129dc9ca)

nefeline commented on tests/e2e/specs/backend/product-query/advanced-filters.test.ts:
```diff
@@ -0,0 +1,160 @@
+				);
+			} );
+
+			it( 'Works on the front end', async () => {
```
What do you think about adding a more descriptive test name here, so we know what exactly to expect on the front-end?

dinhtungdu replied:
Addressed in [8afad60](https://github.com/woocommerce/woocommerce-blocks/pull/7749/commits/8afad60be424aabc3361928d62bbea5c129dc9ca)

nefeline commented on tests/e2e/specs/backend/product-query/advanced-filters.test.ts:
```diff
@@ -0,0 +1,160 @@
+				await shopper.block.goToBlockPage( block.name );
+				expect( await getFrontEndProducts() ).toHaveLength(
+					outOfStockCount
+				);
```
Similar to my previous comment, I believe it would be beneficial to separate these following the same pattern:
```
Enable the filter by stock with the 'Stock status' and 'out of stock' options
- Test the expected behavior in the editor
- Test the expected behavior on the FE

Disable the filter by stock for 'Stock status' and 'out of stock' options
- Test the expected behavior in the editor
- Test the expected behavior on the FE
```

dinhtungdu replied:
Addressed in [8afad60](https://github.com/woocommerce/woocommerce-blocks/pull/7749/commits/8afad60be424aabc3361928d62bbea5c129dc9ca)

nefeline commented on tests/e2e/specs/backend/product-query/common.ts:
```diff
@@ -0,0 +1,267 @@
+	return await findSidebarPanelWithTitle( 'Popular Filters' );
+};
+
+export const getCurrentPopularFilter = async (
```
There are a few utilities here that are specific to certain filter tests, such as this one, `getPopularFilterPanel` and the`selectPopularFilter`: I would recommend moving those to where they are used instead of keeping them in this common file if they won't be shared/used elsewhere.

dinhtungdu replied:
Indeed, they're not common, thanks for this!

nefeline commented on tests/e2e/specs/backend/product-query/advanced-filters.test.ts:
```diff
@@ -0,0 +1,160 @@
+import {
+	setCheckbox,
+	unsetCheckbox,
+	ensureSidebarOpened,
```
Is there a particular reason for us to be importing `ensureSidebarOpened` from `'@woocommerce/e2e-utils'` over here while at the same time importing the same `ensureSidebarOpened` from `'@wordpress/e2e-test-utils'` for the Popular Filter [on common](https://github.com/woocommerce/woocommerce-blocks/blob/3638cd21b16647b5d0045da154949f914b5f5ea1/tests/e2e/specs/backend/product-query/common.ts#L18)? 

Initially, the tests for the advanced filters were failing locally as a consequence of an incorrect selector for the editor when calling `ensureSidebarOpened` (` Error: failed to find element matching selector ".edit-post-sidebar"`). Importing `ensureSidebarOpened` from `'@wordpress/e2e-test-utils'` instead has solved the problem over here:

`import { ensureSidebarOpened } from '@wordpress/e2e-test-utils';`

It looks like there are other tests in our repo, such as `filter-products-by-attribute.test.ts,` also importing from `'@wordpress/e2e-test-utils'`. I believe it would be interesting to aim for consistency here; what do you think?

dinhtungdu replied:
> Is there a particular reason for us to be importing `ensureSidebarOpened` from `'@woocommerce/e2e-utils'` over here while at the same time importing the same `ensureSidebarOpened` from `'@wordpress/e2e-test-utils'` for the Popular Filter [on common](https://github.com/woocommerce/woocommerce-blocks/blob/3638cd21b16647b5d0045da154949f914b5f5ea1/tests/e2e/specs/backend/product-query/common.ts#L18)?

Great catch, it's my mistake, we should use `ensureSidebarOpened` from `@wordpress/e2e-test-utils`. Updated.

nefeline reviewed (changes requested):
Awesome to see these tests in place for the Advanced and Popular Filters, great work! I've added a few recommendations, mostly regarding code reuse and suggestions for the way the tests are organized, so they are more atomic/single-purposed.

Regarding the advanced filters tests, I've added a question and shared some insights [over here](https://github.com/woocommerce/woocommerce-blocks/pull/7749/#discussion_r1059790614).

When running the tests locally for `popular-filters.test.ts`, all of them are consistently passing, except for `Best Selling > Editor preview and block frontend display the same products`: `previewProducts` and  `frontEndProducts` are not exactly the same.

dinhtungdu committed:
use jest.each to avoid duplicated code

dinhtungdu committed:
Use ensureSidebarOpened from @wordpress/e2e-test-utils

dinhtungdu committed:
move utilities for popular filter tests specifically to the test file.

dinhtungdu committed:
more explicit tests for on sale and stock status

dinhtungdu committed:
test default state once

dinhtungdu commented:
Thank you so much for your reviews! @sunyatasattva @nefeline. I addressed your reviews, can you please take another look at this?

> When running the tests locally for `popular-filters.test.ts`, all of them are consistently passing, except for `Best Selling > Editor preview and block frontend display the same products`: `previewProducts` and `frontEndProducts` are not exactly the same.

I have the same issue running the test locally but can't reproduce it on the CI. Somehow, the tests are only flaky in the local machine, not in CI.

sunyatasattva reviewed (approved):
I like the changes! Thanks for refactoring away the duplicated code :)

nefeline reviewed (approved):
Thanks for the updates, great work 🚀 !! All tests are passing here, except for the 'Product Query > Popular Filters › Popular Filters is expanded by default': interesting that it doesn't fail on the CI: one more to the hall of fame of flaky tests.
